### PR TITLE
Message editing

### PR DESCRIPTION
**Pull request by aforge**
Appears to be on par with Android.

### CODE REVIEW

File: TinodeSDK/Topic.swift [hunk 1]
```diff
@@ -1332,6 +1333,25 @@ open class Topic<DP: Codable & Mergeable, DR: Codable & Mergeable, SP: Codable,
         return PromisedReply<ServerMessage>(error: TinodeError.notConnected("Tinode not connected."))
     }
 
+    private func delMessages(inRanges ranges: [MsgRange], hard: Bool) -> PromisedReply<ServerMessage> {
+        store?.msgMarkToDelete(topic: self, ranges: ranges, markAsHard: hard)
+        if attached {
+            return tinode!.delMessage(topicName: self.name, ranges: ranges, hard: hard).then(
+                onSuccess: { [weak self] msg in
+                    if let s = self, let delId = msg?.ctrl?.getIntParam(for: "del"), delId > 0 {
+                        s.clear = delId
```
Comment (or-else) [inline TinodeSDK/Topic.swift:1342]: Check if `s.clear < delId`?

File: TinodeSDK/Topic.swift [hunk 2]
```diff
@@ -1332,6 +1333,25 @@ open class Topic<DP: Codable & Mergeable, DR: Codable & Mergeable, SP: Codable,
...
+                        s.maxDel = delId
```
Comment (or-else) [inline TinodeSDK/Topic.swift:1343]: Check if `s. maxDel < delId`?

File: TinodeSDK/Topic.swift [hunk 3]
```diff
@@ -1386,7 +1413,7 @@ open class Topic<DP: Codable & Mergeable, DR: Codable & Mergeable, SP: Codable,
         let limit = newSeq - description.getSeq
```
Comment (or-else) [inline TinodeSDK/Topic.swift:1413]: Do you still need it? You no longer use it below in line 1416.

File: TinodeSDK/Topic.swift [hunk 4]
```diff
@@ -1386,7 +1413,7 @@ open class Topic<DP: Codable & Mergeable, DR: Codable & Mergeable, SP: Codable,
...
         self.setSeq(seq: newSeq)
         if !self.attached {
-            self.subscribe(set: nil, get: self.metaGetBuilder().withLaterData(limit: limit).build()).thenApply({ _ in
```
Comment (or-else) [inline TinodeSDK/Topic.swift:1389]: Why remove the limit?

Comment (or-else) [reply TinodeSDK/Topic.swift:1389]: Or maybe use a hardcoded limit?

File: Tinodios/MessageViewController+MessageCellDelegate.swift [hunk 5]
```diff
@@ -114,6 +114,20 @@ extension MessageViewController: MessageCellDelegate {
         if !cell.isDeleted, let msgIndex = messageSeqIdIndex[cell.seqId], messages[msgIndex].isSynced {
             menuItems.append(MessageMenuItem(title: NSLocalizedString("Reply", comment: "Menu item"), action: #selector(showReplyPreview(sender:)), seqId: cell.seqId))
             menuItems.append(MessageMenuItem(title: NSLocalizedString("Forward", comment: "Menu item"), action: #selector(showForwardSelector(sender:)), seqId: cell.seqId))
+            let msg = messages[msgIndex]
+            if isFromCurrentSender(message: msg), let content = msg.content {
+                // Only allow editing messages which don't contain certain entity types.
+                var canEdit = true
+                let prohibitedTypes: Set = ["AU", "EX", "FM", "IM", "VC", "VD"]
```
Comment (or-else) [inline Tinodios/MessageViewController+MessageCellDelegate.swift:121]: \+ "QQ"

**Review by or-else — COMMENTED** [review]
Message bubble is not expanded for 'edited':

Screenshot 2022-12-21 at 13 53 57

**Comment by or-else** [comment]
Message view scrolls to the bottom on receiving an edit. It should not move.

**Comment by aforge** [comment]
> Message bubble is not expanded for 'edited':
> 
> Screenshot 2022-12-21 at 13 53 57

Done.

**Comment by aforge** [comment]
> Message view scrolls to the bottom on receiving an edit. It should not move.

Right. Though it just shifts the view somewhat, it doesn't scroll to the bottom.
It appears related to the change in the size of `inputAccessoryView`. I'll fix separately. Thanks for pointing out.

**Pull request merged**